activerecord/lib/active_record/attribute_methods/dirty.rb

```diff
@@ -229,7 +229,13 @@ def attribute_names_for_partial_updates
         end
 
         def attribute_names_for_partial_inserts
-          partial_inserts? ? changed_attribute_names_to_save : attribute_names
+          if partial_inserts?
+            changed_attribute_names_to_save
+          else
+            attribute_names.delete_if do |attr_name|
+              !attribute_changed?(attr_name) && column_for_attribute(attr_name).default_function
+            end
```

Comment:
What about extracting this code into a separate method?

Reply:
Extracted to `changed_attribute_names_without_default_function` but can use better name.